### ISSUE TITLE
Unsound `transmute` in `slint_interpreter_value_to_struct`

### ISSUE THREAD

**Issue by shinmao** [slint-ui/slint#2859]
https://github.com/slint-ui/slint/blob/e1d6f60a70fb887d4943475dd5fe320f2f18620e/internal/interpreter/ffi.rs#L195-L200
The function above used `transmute` to convert `&Struct` to `&StructOpaque`. Take a look at the definition of `Struct`:
https://github.com/slint-ui/slint/blob/e1d6f60a70fb887d4943475dd5fe320f2f18620e/internal/compiler/generator/cpp.rs#L116-L121
I think we need to use `repr(C)` to guarantee the layout of `Struct` to avoid copy on padding bytes.

Please let me know if I am wrong.

**Comment by tronical** [comment]
This is the definition of `StructOpaque`:

https://github.com/slint-ui/slint/blob/be11cb0770a96a9bcb8d2b0627e329dc24aab751/internal/interpreter/ffi.rs#L210

This is the definition of `Struct`:

https://github.com/slint-ui/slint/blob/e1d6f60a70fb887d4943475dd5fe320f2f18620e/internal/interpreter/api.rs#L398

We have an assertion under `StructOpaque` that tries to ensure that there is enough space.

What do you think?

**Comment by shinmao** [comment]
I think you are talking about this assertion, right?
https://github.com/slint-ui/slint/blob/be11cb0770a96a9bcb8d2b0627e329dc24aab751/internal/interpreter/ffi.rs#L216-L217
However, the problem is that the struct of `repr(Rust)` doesn't have stable layout. For example, padding bytes might be added at the beginning of the struct, `transmute` it will lead to uninitialized memory read. With `repr(Rust)`, we are never able to predict the layout of type.
> Also, `transmute` on reference requires caller to take the responsibility.

Therefore, I think your assertion can not check whether the Src type and Dst type are properly aligned.

We can also send a PR of advisories for reviewing this issue. How do you think about it?

**Comment by ogoffart** [comment]
Thanks for taking time to filling a bug report.
 - As @tronical said, the `Struct` being transmuted is the one defined in api.rs
 - The transmute is transmuting references, not the struct itself, so no copy of eventual padding bytes is hapenning.
 - Even if there was padding bytes, it is still safe to transmute between padding bytes. 

The thing is that OpaqueStruct is never constructed or used directly in Rust anyway, it's just an alias for the actual `Struct` that is being used in C++.

Technically, since `HashMap` is not repr(c), it's size and alignment are not stable and could break in a future version, but since we have asserts in place we'd notice that.

**Comment by ogoffart** [comment]
> Therefore, I think your assertion can not check whether the Src type and Dst type are properly aligned.

Yes, that's true, even though usize is the highest alignment, i added now an assert for the alignement.

> We can also send a PR of advisories for reviewing this issue. How do you think about it?

I'd be happy to have review of our unsafe code, yes.

**Comment by shinmao** [comment]
Hi, I thought that I made some mistake in my statement.
> The transmute is transmuting references, not the struct itself, so no copy of eventual padding bytes is happening.

Yes.`StructOpaque` used representation of C; therefore, we don't need to worry about constructing invalid value such as dangling reference. 

> Yes, that's true, even though usize is the highest alignment, i added now an assert for the alignment.

I just took a look at your commit about assertion. Are you sure the function `layout_of` is in module `std::mem`? I could only find the series of `align_of` function. Additionally, I think making sure that the alignment of `Struct` is the multiple of `StructOpaque` would be more precise.

> I'd be happy to have review of our unsafe code, yes.

Sure. I will send the PR to have someone else help us review the unsoundness in `transmute` here.

**Comment by ogoffart** [comment]
> I just took a look at your commit about assertion. Are you sure the function layout_of is in module std::mem? 

Yeah, there was a typo. I had fixed it but somehow forgot to commit before push :-s

**Comment by shinmao** [comment]
RUSTSEC PR: https://github.com/rustsec/advisory-db/pull/1710

**Comment by shinmao** [comment]
@ogoffart due to the fact that `slint_interpreter_value_to_struct` is FFI function, the unsoundness does not lie in this function. I think the only unsound `transmute` we need to take care of is the rust methods where you add the layout check (`size_of` and `align_of`). Therefore, I would change the topic of this issue:p

Btw, I am thinking about whether we should change `size_of::<Src>() - size_of::<Dst>` to `size_of::<Src>() == size_of::<Dst>`. Do you think it would be better?

**Comment by ogoffart** [comment]
I changed the check to check the actual size is equal. I don't think this matters because the Opaque structure is never deferenced anyway.
But I realize that this function could be used from rust to get a reference to such a structure. So I replaced it to return a raw pointer now.

This is only used for ffi from C++ and the C++ also don't de-reference the opaque structure, so this shouldn't lead to any UB.
